### PR TITLE
Svelte CSF Stories

### PR DESCRIPTION
**Pull request by j3rem1e**
Implementation of a svelte story format

### CODE REVIEW

**Review by shilman — APPROVED** [review]
This is tremendous @j3rem1e -- I'm thrilled you were able to get it working without modifying Storybook core 🙌

**Comment by Anthropic** [comment]
@j3rem1e would it be possible, based on your knowledge gained from working on this epically awesome PR here, to actually include a storybook definition within a svelte component in future? Just curious, not even a suggestion.

I admit the documentation part of this PR had me initially thinking that would be possible, till I read it more carefully.

@j3rem1e you are a champion, can't wait to use this.

**Comment by j3rem1e** [comment]
@Anthropic I don't think it will be easy to do that. maybe with a preprocessor, but mixin a component and his stories is probably not a good idea : 
* Story or Meta component should be included only in a Storybook, not in the final prod bundle
* script tag should not be shared either, otherwise test code will be included in the prod bundle

File: package.json [hunk 1]
```diff
@@ -22,39 +19,62 @@
...
+    "@storybook/components": "^6.1.20",
+    "@storybook/core-events": "^6.1.20",
+    "@storybook/svelte": "^6.1.20",
+    "@storybook/theming": "^6.1.20",
     "react": "^16.8.0 || ^17.0.0",
-    "react-dom": "^16.8.0 || ^17.0.0"
+    "react-dom": "^16.8.0 || ^17.0.0",
+    "svelte": "^3.32.3",
+    "svelte-loader": "^2.9.1",
+    "ts-dedent": "^2.0.0"
```
Comment (phated) [inline package.json:72]: `ts-dedent` -> regular dep instead of peer?

File: src/parser/collect-stories.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      },
+    });
+    context.$destroy();
+  } catch (e) {
+    logger.error(`Error extracting stories ${e.toString()}`, e);
+  }
+
+  const { meta } = repositories;
+  if (!meta) {
+    logger.error('Missing <Meta/> tag');
```
Comment (phated) [inline src/parser/collect-stories.js:46]: Is this only a warn-able offense or should this error?

File: src/parser/collect-stories.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    .filter((story) => story.isTemplate)
+    .map((story) => story.id);
+
+  // check for duplicate templates
+  const duplicateTemplatesId = templatesId.filter(
+    (item, index) => templatesId.indexOf(item) !== index
+  );
+
+  if (duplicateTemplatesId.length > 0) {
+    logger.warn(
```
Comment (phated) [inline src/parser/collect-stories.js:63]: Is this only a warn-able offense or should this error?

File: src/parser/collect-stories.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+  // collect templates id
+  const templatesId = repositories.stories
+    .filter((story) => story.isTemplate)
+    .map((story) => story.id);
+
+  // check for duplicate templates
+  const duplicateTemplatesId = templatesId.filter(
+    (item, index) => templatesId.indexOf(item) !== index
+  );
```
Comment (phated) [inline src/parser/collect-stories.js:60]: Thoughts on using `(new Set(templatesId)).size === templatesId.length`

File: src/parser/collect-stories.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      .filter((story) => !story.isTemplate)
+      .reduce((all, story) => {
+        const { id, name, template, component, source = false, ...props } = story;
+
+        const storyId = extractId(story);
+        if (!storyId) {
+          return all;
+        }
+
+        const unknowTemplate = template != null && templatesId.indexOf(template) < 0;
```
Comment (phated) [inline src/parser/collect-stories.js:80]:
```suggestion
        const unknownTemplate = template != null && templatesId.indexOf(template) < 0;
```
(and references below)

File: src/parser/extract-id.test.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import { extractId } from './extract-id';
+
+describe('extract-id', () => {
+  test('name with spaces', () => {
+    expect(extractId({ name: 'Name with spaces' })).toBe('Name_with_spaces');
```
Comment (phated) [inline src/parser/extract-id.test.js:5]: Is underscore the convention or just preference?

File: src/parser/extract-stories.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+ * @returns Map of storyName -> source
+ */
+export function extractStories(component: string): Record<string, StoryDef> {
+  // compile
+  const { ast } = svelte.compile(component);
+
+  const stories: Record<string, StoryDef> = {};
+  svelte.walk(ast.html, {
+    enter(node: any) {
+      if (node.type === 'InlineComponent' && (node.name === 'Story' || node.name === 'Template')) {
```
Comment (phated) [inline src/parser/extract-stories.ts:43]: What happens if someone creates their own component named `Story` or `Template`? Will those match here?

File: src/parser/svelte-stories-loader.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+  const componentName = getNameFromFilename(resource);
+
+  const source = readFileSync(resource).toString();
+
+  const stories = extractStories(source);
+
+  const storyDef = Object.entries(stories)
+    .filter(([, def]) => !def.template)
+    .map(([id]) => `export const ${id} = __storiesMetaData.stories[${JSON.stringify(id)}]`)
```
Comment (phated) [inline src/parser/svelte-stories-loader.ts:51]: So the collected stories are all available in the `__storiesMetaData` object, but their story IDs are statically exported as mappings to the object reference?

File: src/parser/svelte-stories-loader.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    .join('\n');
+
+  const codeWithoutDefaultExport = code.replace('export default ', '//export default');
```
Comment (phated) [inline src/parser/svelte-stories-loader.ts:54]: What's this for?

File: src/parser/svelte-stories-loader.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+  return dedent`${codeWithoutDefaultExport}
+    const { default: parser } = require('${parser}');
+    const __storiesMetaData = parser(${componentName}, ${JSON.stringify(stories)});
```
Comment (phated) [inline src/parser/svelte-stories-loader.ts:58]: So the parser is still run dynamically, even though the stories were already parsed?

**Review by phated — COMMENTED** [review]
Had a few questions about how this works 😄

File: src/parser/collect-stories.js (hunk 2 above)
Comment (j3rem1e) [reply src/parser/collect-stories.js:46]: This is an error, the whole file will be ignored, the same as a stories.js file without a default export.

I didn't throw an Error to not prevent storybook to launch.

File: src/parser/collect-stories.js (hunk 3 above)
Comment (j3rem1e) [reply src/parser/collect-stories.js:63]: Just a warning, the duplicate templates will not be used.

File: src/parser/collect-stories.js (hunk 4 above)
Comment (j3rem1e) [reply src/parser/collect-stories.js:60]: I use the duplicates id in the warning below

File: src/parser/extract-id.test.js (hunk 6 above)
Comment (j3rem1e) [reply src/parser/extract-id.test.js:5]: A preference, I have to generate a valid js id from a story name.

File: src/parser/extract-stories.ts (hunk 7 above)
Comment (j3rem1e) [reply src/parser/extract-stories.ts:43]: You're right, If a user write : 

```html
<script>
  import { Story as SBStory } from '@storybook/addon-svelte-csf'
  import Story from './Story.svelte'
</script>
``` 

then this check will fail. I can try to parse the import statements in the script tag.

File: src/parser/svelte-stories-loader.ts (hunk 8 above)
Comment (j3rem1e) [reply src/parser/svelte-stories-loader.ts:51]: Yes, this is how I can write this addon without changing storybook/core. I generate static exports by parsing the AST, and this export reference an object generated by executing the <Story/> in a 'collect without rendering' context.

File: src/parser/svelte-stories-loader.ts (hunk 9 above)
Comment (j3rem1e) [reply src/parser/svelte-stories-loader.ts:54]: In svelte, the component are exported as a default export. this is not valid in CSF, the default export should be the <Meta/> tag.

This line is just a 'cheap' way to transform the js generated by svelte by removing this default export.

File: src/parser/svelte-stories-loader.ts (hunk 10 above)
Comment (j3rem1e) [reply src/parser/svelte-stories-loader.ts:58]: Yes, the Stories are executed in a context which disable rendering, but collect every args/loaders/parameters

File: package.json (hunk 1 above)
Comment (j3rem1e) [reply package.json:72]: I'll fix that

File: src/parser/extract-id.test.js (hunk 6 above)
Comment (phated) [reply src/parser/extract-id.test.js:5]: I wonder if `NameWithSpaces` would be the more common way to write a story ID

**Comment by betaboon** [comment]
i just checked out this branch and ran `npm run build` followed by `npm run storybook` and ran into the following error for all svelte-stories:
```
WARNING in ./stories/button.stories.svelte
Module build failed (from ./dist/parser/svelte-stories-loader.js):
TypeError: name.replaceAll is not a function
    at extractId (/home/betaboon/src/js/addon-svelte-csf/dist/parser/extract-id.js:17:15)
    at Object.enter (/home/betaboon/src/js/addon-svelte-csf/dist/parser/extract-stories.js:86:43)
    at visit (/home/betaboon/src/js/addon-svelte-csf/node_modules/svelte/compiler.js:5256:11)
    at visit (/home/betaboon/src/js/addon-svelte-csf/node_modules/svelte/compiler.js:5288:13)
    at Object.walk (/home/betaboon/src/js/addon-svelte-csf/node_modules/svelte/compiler.js:5207:10)
    at extractStories (/home/betaboon/src/js/addon-svelte-csf/dist/parser/extract-stories.js:72:10)
    at Object.transformSvelteStories (/home/betaboon/src/js/addon-svelte-csf/dist/parser/svelte-stories-loader.js:64:52)
```
it is working for me with the following patch:
```diff
diff --git a/src/parser/extract-id.ts b/src/parser/extract-id.ts
index e47f87a..cb54669 100644
--- a/src/parser/extract-id.ts
+++ b/src/parser/extract-id.ts
@@ -4,5 +4,5 @@ export function extractId({ id, name }: { id?: string; name?: string }): string
     return id;
   }
 
-  return name.replaceAll(/[^a-zA-Z0-9_]/g, '_');
+  return name.replace(/[^a-zA-Z0-9_]/g, '_');
 }

```

**Comment by j3rem1e** [comment]
@betaboon this should be fixed, I removed `replaceAll` which is not supported by all platforms.

**Comment by betaboon** [comment]
> @betaboon this should be fixed, I removed `replaceAll` which is not supported by all platforms.

i can confirm that it is fixed (after the force-push) :) thank you

**Comment by betaboon** [comment]
just fyi: using `argTypes` works great as well:

```svelte
<script>
  import { Meta, Story, Template } from "@storybook/addon-svelte-csf";
  import Button from "./Button.svelte";

  const meta = {
    title: "Example/Button",
    component: Button,
    argTypes: {
      backgroundColor: { control: "color" },
      size: {
        control: {
          type: "inline-radio",
          options: ["small", "medium", "large"],
        },
      },
      onClick: { action: "onClick" },
    },
  };
</script>

<Meta {...meta} />

<Template let:args>
  <Button {...args} on:click={args.onClick} />
</Template>

<Story name="Primary" args={{ label: 'Button', primary: true }} />

<Story name="Secondary" args={{ label: 'Button' }} />

<Story name="Large" args={{ label: 'Button', size: 'large' }} />

<Story name="Small" args={{ label: 'Button', size: 'small' }} />

```

PS: would it be possible to get a pre-release on npm ? :)

**Review by phated — APPROVED** [review]
@shilman I'm inclined to say this is good-to-go unless you have any feedback on my couple of comments.

**Comment by shilman** [comment]
Sorry just seeing the notification now--will review in the morning 🙏

File: src/jest-transform.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+const svelte = require('svelte/compiler');
+
+const parser = require.resolve('./parser/collect-stories').replace(/[/\\]/g, '/');
```
Comment (hkievet) [inline src/jest-transform.js:3]: Why `[/\\]` and not just `[\\]`?

**Pull request merged**